### PR TITLE
Clear Undefined Behavior sanitizer findings (GH #8426)

### PR DESCRIPTION
**Pull request by noloader**
Clear Undefined Behavior sanitizer findings

### CODE REVIEW

File: crypto/lhash/lhash.c [hunk 1]
```diff
@@ -463,7 +463,7 @@ unsigned long lh_strhash(const char *c)
         v = n | (*c);
         n += 0x100;
         r = (int)((v >> 2) ^ v) & 0x0f;
-        ret = (ret << r) | (ret >> (32 - r));
+        ret = (ret << (r&31)) | (ret >> ((32-r)&31));
```
Comment (t-j-h) [inline crypto/lhash/lhash.c]: Not protected by PEDANTIC and also not in master ...

File: ssl/ssltest.c [hunk 2]
```diff
@@ -470,10 +470,17 @@ static int cb_server_alpn(SSL *s, const unsigned char **out,
 static int verify_alpn(SSL *client, SSL *server)
 {
     const unsigned char *client_proto, *server_proto;
+    const unsigned char null_protocol[1] = {0};
```
Comment (t-j-h) [inline ssl/ssltest.c]: Not protected by PENDANTIC and I cannot find an equivalent change in master

File: crypto/ec/ecp_nistp224.c [hunk 3]
```diff
@@ -317,10 +317,22 @@ const EC_METHOD *EC_GFp_nistp224_method(void)
  */
 static void bin28_to_felem(felem out, const u8 in[28])
 {
+#ifdef PEDANTIC
+    uint64_t temp;
```
Comment (mattcaswell) [inline crypto/ec/ecp_nistp224.c]: Nit: please put a blank line after the declaration

File: crypto/ec/ecp_nistp224.c [hunk 4]
```diff
@@ -317,10 +317,22 @@ const EC_METHOD *EC_GFp_nistp224_method(void)
...
+    memcpy(&temp, in+ 0, 8);
```
Comment (mattcaswell) [inline crypto/ec/ecp_nistp224.c]: Nit: please put a space either side of "+". Here and below.

Comment (mattcaswell) [reply crypto/ec/ecp_nistp224.c]: Rather than 8, please use `sizeof(temp)`

File: crypto/ec/ecp_nistp521.c [hunk 5]
```diff
@@ -139,6 +139,27 @@ static const limb bottom58bits = 0x3ffffffffffffff;
  */
 static void bin66_to_felem(felem out, const u8 in[66])
 {
+#ifdef PEDANTIC
+    limb temp;
```
Comment (mattcaswell) [inline crypto/ec/ecp_nistp521.c]: Same comments as above: blank line after declaration, spaces around "+"

File: crypto/ec/ecp_nistp521.c [hunk 6]
```diff
@@ -156,6 +178,46 @@ static void bin66_to_felem(felem out, const u8 in[66])
  */
 static void felem_to_bin66(u8 out[66], const felem in)
 {
+#ifdef PEDANTIC
+    limb temp;
```
Comment (mattcaswell) [inline crypto/ec/ecp_nistp521.c]: nit: blank line after declaration

File: crypto/lhash/lhash.c (hunk 1 above)
Comment (mattcaswell) [reply crypto/lhash/lhash.c]: Nit: add space around "&" and "-"

File: crypto/cast/cast_lcl.h [hunk 7]
```diff
@@ -153,7 +153,7 @@
 #if defined(OPENSSL_SYS_WIN32) && defined(_MSC_VER)
 # define ROTL(a,n)     (_lrotl(a,n))
 #elif defined(PEDANTIC)
-# define ROTL(a,n)     ((((a)<<(n))&0xffffffffL)|((a)>>((32-(n))&31)))
+# define ROTL(a,n)     ((((a)<<((n)&31))|((a)>>((32-(n))&31)))&0xffffffffL)
```
Comment (kroeckx) [inline crypto/cast/cast_lcl.h]: Is there a reason why this is different than what we have in the master branch? There we have:
```
# define ROTL(a,n)     ((((a)<<(n))&0xffffffffL)|((a)>>((32-(n))&31)))
```

File: crypto/lhash/lhash.c (hunk 1 above)
Comment (kroeckx) [reply crypto/lhash/lhash.c]: The line above does & 0x0f, there is little point in doing r&31.

Comment (kroeckx) [reply crypto/lhash/lhash.c]: I also don't see the point in adding pedantic and non-pedantic version of this.

File: ssl/ssltest.c (hunk 2 above)
Comment (kroeckx) [reply ssl/ssltest.c]: In master the code has been changed to:
```
   if (client_proto != NULL &&
        memcmp(client_proto, server_proto, client_proto_len) != 0) {
        BIO_printf(bio_stdout, "ALPN selected protocols differ!\n");
        goto err;
    }
```

File: ssl/t1_lib.c [hunk 8]
```diff
@@ -133,7 +133,8 @@ int ssl_check_serverhello_tlsext(SSL *s);
 #endif
 
 #define CHECKLEN(curr, val, limit) \
-    (((curr) >= (limit)) || (size_t)((limit) - (curr)) < (size_t)(val))
+    (((size_t)(curr) >= (size_t)(limit)) || \
+     ((size_t)(limit) - (size_t)(curr)) < (size_t)(val))
```
Comment (kroeckx) [inline ssl/t1_lib.c]: Casting pointers to a size_t is wrong, but happens to work on most platforms. If you need to cast it, it needs to be an uintptr_t. But I currently don't see what is wrong with the old macro.

Note that that code is also still present in 1.1.0, but was removed in 1.1.1.

**Comment by noloader** [comment]
On Wed, Apr 24, 2019 at 1:16 PM Kurt Roeckx <notifications@github.com>
wrote:

> *@kroeckx* commented on this pull request.
> ------------------------------
>
> In ssl/t1_lib.c
> <https://github.com/openssl/openssl/pull/8819#discussion_r278231600>:
>
> > @@ -133,7 +133,8 @@ int ssl_check_serverhello_tlsext(SSL *s);
>  #endif
>
>  #define CHECKLEN(curr, val, limit) \
> -    (((curr) >= (limit)) || (size_t)((limit) - (curr)) < (size_t)(val))
> +    (((size_t)(curr) >= (size_t)(limit)) || \
> +     ((size_t)(limit) - (size_t)(curr)) < (size_t)(val))
>
> Casting pointers to a size_t is wrong, but happens to work on most
> platforms. If you need to cast it, it needs to be an uintptr_t. But I
> currently don't see what is wrong with the old macro.
>
Yeah, I was not sure how OpenSSL handled uintptr_t in C89.

size_t will be safe except for segmented memory, like i286/i386. If you are
still supporting DOS then it should probably change.

Here is the finding on Aarch64 Tinkerboard (
https://www.asus.com/us/Single-Board-Computer/Tinker-Board/). Debian 9.8
Stretch, GCC 6.3.

test sslv2/sslv3 with server authentication
server authentication
t1_lib.c:1372:13: runtime error: signed integer overflow: -2147468736 -
2147482398 cannot be represented in type 'int [28]'
t1_lib.c:1400:13: runtime error: signed integer overflow: -2147468736 -
2147482406 cannot be represented in type 'int [28]'
t1_lib.c:1455:13: runtime error: signed integer overflow: -2147468736 -
2147482438 cannot be represented in type 'int [28]'
t1_lib.c:1543:9: runtime error: signed integer overflow: -2147468736 -
2147482474 cannot be represented in type 'int [28]'

The source code lines refer to CHECKLEN.

Jeff

**Comment by kroeckx** [comment]
So you do (char *)0x80003a40 - (char *)0x7ffffb1e = 0x3f22. As long as those are valid pointers pointing to the same object (or 1 past it), this is be defined behaviour. Subtracting 2 pointers gives you a ptrdiff_t, which is a signed type, but as long as the size is smaller than 2^31 there is no undefined behaviour. I think this is a compiler bug.

I assume that you're running the 32 bit armhf (or armel) port, and not the arm64 port. But I wonder why we can have valid pointers > 0x80000000 there. Or is this a 64 bit kernel and 32 bit userland?

**Comment by noloader** [comment]
On Wed, Apr 24, 2019 at 5:16 PM Kurt Roeckx <notifications@github.com> wrote:
>
> So you do (char *)0x80003a40 - (char *)0x7ffffb1e = 0x3f22. As long as those are valid pointers pointing to the same object (or 1 past it), this is be defined behaviour. Subtracting 2 pointers gives you a ptrdiff_t, which is a signed type, but as long as the size is smaller than 2^31 there is no undefined behaviour. I think this is a compiler bug.
>
> I assume that you're running the 32 bit armhf (or armel) port, and not the arm64 port. But I wonder why we can have valid pointers > 0x80000000 there. Or is this a 64 bit kernel and 32 bit userland?

Oh, you're right, my bad. It is armv7l, not Aarch64. (I've been
testing on 6 machines concurrently, and I crossed my wires).

Jeff

**Comment by mspncp** [comment]
Hi @noloader,

this pull request was marked as 'resolved: wont fix', because OpenSSL 1.0.2 has reached end-of-life. Please let us know whether this change might be still relevant for the next OpenSSL 3.0 release (and OpenSSL 1.1.1 LTS if it's a bugfix) and consider raising a new new pull request against the master branch if you think that's the case.

**Comment by levitte** [comment]
There's been no further comment, so closing.  Please raise a new PR if this needs work in current branches (1.1.1 and master)

**Pull request closed without merging**